File: lexer.c
#include "lexer.h"

enum lexerr_type {
    LEX_SUCCEEDED,
    UNEXPECTED_CHAR,
    INCORRECT_INDENT,
};

#define SUCCESS ((struct lexerr) { .type = LEX_SUCCEEDED })
#define MK_TOK(t) ((struct token) { .type = t })
#define MK_LEXERR(e,s) ((struct lexerr) { .type = e, .loc = (s).loc })

struct state {
    int loc;
    struct indent_list* indents;
};

struct indent_list {
    int indent;
    struct indent_list* next;
};

void indent(struct state* state, int amount) {
    struct indent_list* new = malloc(sizeof(struct indent_list));
    new->indent = amount;
    new->next = state->indents;
    state->indents = new;
}

void dedent(struct state* state) {
    struct indent_list* new = state->indents->next;
    free(state->indents);
    state->indents = new;
}

int get_indent(struct state* state) {
    if (state->indents == NULL)
        return 0;
    else
        return state->indents->indent;
}

int get_last_indent(struct state* state) {
    if (state->indents == NULL)
        return -1;
    else if (state->indents->next == NULL)
        return 0;
    else
        return state->indents->next->indent;
}

struct lexerr lex_ident(char* str, struct state* state, struct token* ret) {
    int i = state->loc;
    for (; isalpha(str[i]) || isdigit(str[i]) || str[i] == '_'; ++i);

    int len = i - state->loc;
    char* ident = malloc(sizeof(char) * (len+1));
    memcpy(ident, str+state->loc, len);
    ident[len] = '\0';

    state->loc = i;
    ret->type = IDENT_TOK;
    ret->ident = ident;
    return SUCCESS;
}

struct lexerr lex_number(char* str, struct state* state, struct token* ret) {
    int i;
    int iacc = 0;

    for (i = state->loc; isdigit(str[i]); ++i) {
        iacc = iacc * 10 + (str[i] - '0');
    }

    if (str[i] == '.') {
        double racc = 0;
        double exp = 0.1;

        while (isdigit(str[i])) {
            racc += exp * (str[i]);
            exp /= 10;
        }
        
        state->loc = i;
        ret->type = REAL_TOK;
        ret->real = racc;
        return SUCCESS;
    } else {
        state->loc = i;
        ret->type = INT_TOK;
        ret->intv = iacc;
        return SUCCESS;
    }
}

struct lexerr lex_token(char* str, struct state* state, struct token* ret) {
    if (str[state->loc] == '\0') {
        ret->type = END_TOK;
        return SUCCESS;
    } else if (str[state->loc] == '(') {
        ++state->loc;
        ret->type = LPAREN_TOK;
        return SUCCESS;
    } else if (str[state->loc] == ')') {
        ++state->loc;
        ret->type = RPAREN_TOK;
        return SUCCESS;
    } else if (isalpha(str[state->loc]) || str[state->loc] == '_') {
        return lex_ident(str, state, ret);
    } else if (isdigit(str[state->loc])) {
        return lex_number(str, state, ret);
    } else {
        return MK_LEXERR(UNEXPECTED_CHAR, *state);
    }
}

struct lexerr lex(char* str, struct token** ret) {
    struct state state = { .loc = 0, .indents = NULL };
    struct token tok;
    struct token* toks = malloc(sizeof(struct token));
    struct lexerr err;
    int num_toks = 0;

    for (;;) {
        if (str[state.loc] == '\n') {
            int new_indent = 0;
            int i;
            for (i = state.loc + 1; str[i] == ' '; ++i, ++new_indent);
            state.loc = i;
            
            int last_indent = get_last_indent(&state);
            if (last_indent != -1 && new_indent <= last_indent) {
                int num_dedents = 0;
                struct indent_list* iter = state.indents;
                for (; iter != NULL && iter->indent != new_indent; iter = iter->next, ++num_dedents);

                if (iter != NULL || (iter == NULL && new_indent == 0)) {
                    for (int i = 0; i < num_dedents; ++i) {
                        dedent(&state);
                        toks = realloc(toks, sizeof(struct token) * ++num_toks);
                        toks[num_toks-1].type = DEDENT_TOK;
                    } 
                } else {
                    return MK_LEXERR(INCORRECT_INDENT, state);
                }
            } else if (new_indent > get_indent(&state)) {
                indent(&state, new_indent);
                toks = realloc(toks, sizeof(struct token) * ++num_toks);
                toks[num_toks-1].type = INDENT_TOK; 
            } else if (new_indent == get_indent(&state)) {
                toks = realloc(toks, sizeof(struct token) * ++num_toks);
                toks[num_toks-1].type = NEWLINE_TOK;
            } else {
                printf("This shouldn't happen...\n");
            }
        }

        while (isspace(str[state.loc])) ++state.loc;

        err = lex_token(str, &state, &tok);
        if (!LEX_SUCCESSFUL(err)) {
            return err;
        } else {
            toks = realloc(toks, sizeof(struct token) * ++num_toks);
            toks[num_toks-1] = tok;

            if (tok.type == END_TOK) {
                *ret = toks;
                return SUCCESS;
            }
        }
    }
}

#undef t
#define t(x) #x
char* token2str_arr[] = {
    TOKEN_HACK
};

void display_token(struct token token) {
    printf("%s", token2str_arr[token.type]);
    
    switch (token.type) {
        case IDENT_TOK:
            printf(" %s", token.ident);
            break;
        case INT_TOK: 
            printf(" %d", token.intv);
            break;
        case REAL_TOK:
            printf(" %f", token.real);
            break;
        default: break;
    }
}

void display_lexerr(char* str, struct lexerr err) {
    switch (err.type) {
        case LEX_SUCCEEDED:
            break;
        case UNEXPECTED_CHAR:
            fprintf(stderr, "unexpected char %c at %d\n", str[err.loc], err.loc);
            break;
        case INCORRECT_INDENT:
            fprintf(stderr, "incorrect indent at %d\n", err.loc);
            break;
    }
}

File: Lexer.hs
{-# LANGUAGE GeneralizedNewtypeDeriving, MultiWayIf #-}

module Lexer where

import Data.Char
import Control.Monad.State
import Control.Monad.Except

data SourcePos = SourcePos 
               { srcLine :: !Int 
               , srcChar :: !Int
               }

instance Show SourcePos where
    show (SourcePos y x) = show y ++ ":" ++ show x

data LexErr = LexErr SourcePos LexErrType

instance Show LexErr where
    show (LexErr sp e) = show sp ++ "\t" ++ show e

data LexErrType = Unexpected Char
                | EndOfInput
                | WrongIndent
                deriving Eq

instance Show LexErrType where
    show e = case e of
        Unexpected c -> "unexpected " ++ [c]
        EndOfInput -> "unexpected end of input"

data Token = Token SourcePos TokType

instance Show Token where
    show (Token _ t) = show t

data TokType = NewlineTok
             | IndentTok
             | DedentTok
             | LParenTok
             | RParenTok
             | IntTok Integer
             | RealTok Double
             | IdentTok String
             deriving Eq

instance Show TokType where
    show t = case t of
        NewlineTok -> "end of file"
        IndentTok -> "indent"
        DedentTok -> "dedent"
        LParenTok -> "left parenthesis"
        RParenTok -> "right parenthesis"
        IntTok n -> "literal integer " ++ show n
        RealTok x -> "literal real " ++ show x
        IdentTok s -> "identifier " ++ s

data LexState = LexState
              { lexInput :: String
              , lexOutput :: [Token]
              , lexLine :: Int
              , lexChar :: Int
              , lexIndents :: [Int]
              } deriving Show

newtype Lexer a = Lexer (StateT LexState (Except LexErr) a)
                deriving ( Functor
                         , Applicative
                         , Monad
                         , MonadState LexState
                         , MonadError LexErr
                         )

runLexer :: Lexer a -> String -> Either LexErr [Token]
runLexer (Lexer l) s = runExcept (reverse . lexOutput <$> execStateT l st)
  where st = LexState
           { lexInput = s
           , lexOutput = []
           , lexLine = 1
           , lexChar = 1
           , lexIndents = []
           }

getPos :: Lexer SourcePos
getPos = do
    l <- gets lexLine
    c <- gets lexChar
    return (SourcePos l c)

throwLexErr :: LexErrType -> Lexer a
throwLexErr e = do
    p <- getPos
    throwError (LexErr p e)

outputAt :: SourcePos -> TokType -> Lexer ()
outputAt p t = modify $ \st -> st { lexOutput = Token p t : lexOutput st }

output :: TokType -> Lexer ()
output t = do
    p <- getPos
    outputAt p t

indent :: Int -> Lexer ()
indent n = do
    output IndentTok
    modify $ \st -> st { lexIndents = n : lexIndents st }

dedent :: Lexer ()
dedent = do
    output DedentTok
    modify $ \st -> st { lexIndents = tail (lexIndents st) }

newline :: Lexer ()
newline = output NewlineTok

-- Very unsafe
peekChar :: Lexer Char
peekChar = do
    (x:_) <- gets lexInput
    return x 

skipChar :: Lexer ()
skipChar = modify $ \st -> st { lexInput = tail (lexInput st) }

skipWhile :: (Char -> Bool) -> Lexer ()
skipWhile cond = do
    eol <- endOfLine
    if eol then return () else do
        c <- peekChar
        if cond c 
          then skipChar *> skipWhile cond
          else return ()

endOfInput :: Lexer Bool
endOfInput = null <$> gets lexInput

endOfLine :: Lexer Bool
endOfLine = do
    eoi <- endOfInput
    if eoi then return True else (== '\n') <$> peekChar

lexTokens :: Lexer ()
lexTokens = do
    eoi <- endOfInput
    if eoi
      then do n <- length <$> gets lexIndents
              replicateM_ n dedent
      else lexToken *> lexTokens

isIdent :: Char -> Bool
isIdent = (`elem` (['a'..'z'] ++ ['A'..'Z'] ++ "_"))

isSpecial :: Char -> Bool
isSpecial = (`elem` "+-*/()[]")

lexToken :: Lexer ()
lexToken = do
    c <- peekChar
    if | c == '\n' -> do
         skipChar
         xs <- gets lexInput
         if all isSpace (takeWhile (/='\n') xs)
           then skipWhile isSpace *> lexToken 
           else lexIndent
       | isIdent c -> lexIdent
       | isDigit c -> lexNumber
       | isSpace c -> skipWhile isSpace
       | otherwise -> throwLexErr (Unexpected c)
     
lexIndent :: Lexer ()
lexIndent = do
    n <- countIndent
    xs <- gets lexIndents
    case xs of
        [] -> if n /= 0 then indent n else newline
        xs -> findIndent n xs

  where countIndent = do
            xs <- gets lexInput
            return (length (takeWhile isSpace xs))
        
        findIndent n (x:xs) = 
            case compare n x of
                EQ -> newline
                GT -> indent n
                _  -> do ds <- findDedents n xs
                         replicateM_ ds dedent
        
        findDedents n [] = return 1
        findDedents n (x:xs) =
            case compare n x of
                EQ -> return 1
                LT -> (+1) <$> findDedents n xs
                _  -> throwLexErr WrongIndent

collect :: (Char -> Bool) -> (Char -> Bool) -> Lexer String
collect contCond stopCond = do
    eol <- endOfLine
    if eol then return "" else do
        c <- peekChar 
        if | contCond c -> (c :) <$> (skipChar *> collect contCond stopCond)
           | stopCond c -> return ""
           | otherwise -> throwLexErr (Unexpected c)
         
lexIdent :: Lexer ()
lexIdent = do
    p <- getPos
    str <- collect (\c -> isIdent c || isDigit c)
                   (\c -> isSpecial c || isSpace c)
    outputAt p (IdentTok str)

lexNumber :: Lexer ()
lexNumber = do
    p <- getPos
    int <- collect isDigit (\c -> isSpecial c || isSpace c || c == '.')
    eol <- endOfLine
    if eol then outputAt p (IntTok (read int)) else do
        c <- peekChar
        if c == '.'
          then do skipChar
                  real <- collect isDigit (\c -> isSpecial c || isSpace c)
                  outputAt p (RealTok (read (int ++ "." ++ real)))
          else outputAt p (IntTok (read int))

